fmn/consumer/consumer.py

```diff
@@ -49,16 +49,33 @@ def __init__(self, *args, **kwargs):
         log.debug("Loading rules from fmn.rules")
         self.valid_paths = fmn.lib.load_rules(root="fmn.rules")
 
-        self.cached_preferences = None
+        # Initialize our in-memory cache of the FMN preferences database
         self.cached_preferences_lock = threading.Lock()
+        self.cached_preferences = None
+        session = self.make_session()
+        self.refresh_cache(session)
+        session.close()
 
         log.debug("FMNConsumer initialized")
 
-    def refresh_cache(self, session, topic=None, msg=None):
-        log.info("Loading and caching preferences")
+    def refresh_cache(self, session, openid=None):
+        log.debug("Acquiring cached_preferences_lock")
         with self.cached_preferences_lock:
-            self.cached_preferences = fmn.lib.load_preferences(
-                session, self.hub.config, self.valid_paths, cull_disabled=True)
+            if not openid or self.cached_preferences is None:
+                log.info("Loading and caching all preferences for all users")
+                self.cached_preferences = fmn.lib.load_preferences(
+                    session, self.hub.config, self.valid_paths,
+                    cull_disabled=True)
+            else:
+                log.info("Loading and caching preferences for %r" % openid)
+                old_preferences = [p for p in self.cached_preferences
+                                   if p['user']['openid'] == openid]
+                new_preferences = fmn.lib.load_preferences(
+                    session, self.hub.config, self.valid_paths,
+                    cull_disabled=True, openid=openid)
+                self.cached_preferences.extend(new_preferences)
```

Comment:
Yeah, I thought about it... but there's a chance for a race condition here.  The ``self.cached_preferences`` list is shared between ``N`` worker threads.

While *this* thread is executing this code and updating the cache of this one person's preferences, the other threads and plugging away on messages and to do that, they use the cached preferences.  If we *remove* the person's prefs first and then *add* the new ones, there's a moment in-between where another thread could try to check if that user should receive a message, but at that moment they have *zero* preferences in the cache.

By adding the new ones first and removing the old ones last, in the worst case there is that moment in-between where the user has two sets of preferences in the list.  The message-matching code will handle that "gracefully" and just send a message if either set of preferences match.

> I am afraid if we get a conflict of identifiers we would remove the new pref we just added.

This won't happen, I think.  We are removing the preferences by value and ``.remove(item)`` only removes the first element it finds:

```python
>>> list = ["thing", "thing", "thing"]
>>> list.remove("thing")
>>> list
['thing', 'thing']
```